lib/Doctrine/ORM/Mapping/EscapingQuoteStrategy.php

```diff
@@ -0,0 +1,176 @@
+<?php
+/*
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license. For more information, see
+ * <http://www.doctrine-project.org>.
+ */
+
+namespace Doctrine\ORM\Mapping;
+
+use Doctrine\DBAL\Platforms\AbstractPlatform;
+
+/**
+ * A set of rules for determining the physical column, alias and table quotes and automatically escape database reserved
+ * keyword
+ *
+ *@author  Xavier HUBERTY <x.huberty@thecodingmachine.com>
+ */
+class EscapingQuoteStrategy implements QuoteStrategy
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getColumnName($fieldName, ClassMetadata $class, AbstractPlatform $platform)
+    {
+        if(isset($class->fieldMappings[$fieldName]['quoted'])){
+            return $platform->quoteIdentifier($class->fieldMappings[$fieldName]['columnName']);
+        }
+        $reservedKeyList = $platform->getReservedKeywordsList();
+        if($reservedKeyList->isKeyword($fieldName)){
+            return $platform->quoteIdentifier($class->fieldMappings[$fieldName]['columnName']);
+        }
+        return $class->fieldMappings[$fieldName]['columnName'];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getTableName(ClassMetadata $class, AbstractPlatform $platform)
+    {
+        if(isset($class->table['quoted'])){
+            return $platform->quoteIdentifier($class->table['name']);
+        }
+        $reservedKeyList = $platform->getReservedKeywordsList();
+        if($reservedKeyList->isKeyword($class->table['name'])){
+            return $platform->quoteIdentifier($class->table['name']);
+        }
+         return $class->table['name'];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSequenceName(array $definition, ClassMetadata $class, AbstractPlatform $platform)
+    {
+        if(isset($definition['quoted'])){
+            return $platform->quoteIdentifier($class->table['name']);
```

Comment:
Hi Ocramius, I sorry to respond so lately but I didn't have much time to work around this PR.

I've implement test for this class but it always fail because the EscapingQuoteStrategy need a real platform in order to get the protected keyword.  For the phpunit test the platform provide does'nt have the method I use for the quoteStrategy. I think that is why the test for the ASCIIQuoteStrategy use the defaultQuoteStrategy instead of the ASCII one.